### DISCUSSION TITLE
How to create and update ui.table using pandas dataframe?

### DISCUSSION THREAD

**Discussion by petteri111** [zauberzeug/nicegui#30] [Q&A] [answered]
Any easy way to use pandas dataframe when creating table? How to auto update it when dataframe changes?

**Comment by falkoschindler** [comment]
Currently NiceGUI itself has no "official" API for pandas dataframes. But since it builds on [JustPy](https://justpy.io/), you can reach into NiceGUI's `table` and access the `table.view`, which is an `AgGrid` providing a method `load_pandas_frame()`:

```python3
df = pd.DataFrame(data={'col1': [1, 2], 'col2': [3, 4]})

table = ui.table({})
table.view.load_pandas_frame(df)
```

When the dataframe changes, you can simply reload it:

```
def increment():
    df['col1'][0] += 1
    table.view.load_pandas_frame(df)

ui.button('increment', on_click=increment)
```

**Reply by petteri111** [reply]
So nice! Thank you!
Placing table to card seems to be somewhat tricky? I was trying to modificate ui.card styles height and width but cannot get it to look ok?

**Reply by falkoschindler** [reply]
@petteri111 Yes, layouting a web application with a combination of CSS, Tailwind, Quasar, JustPy and NiceGUI can be tricky. While the table tends to fill all available space, the card does not grow automatically in some scenarios. You might want to limit the table's maximum width or height and set a card size, e.g. using Tailwind properties:

```python
with ui.card().classes('w-64'):
    table = ui.table({}).classes('max-h-32')
    df = pd.DataFrame(data={'col1': [1, 2], 'col2': [3, 4]})
    table.view.load_pandas_frame(df)
```

Regarding Tailwind: Some elements come with some initial styling. When e.g. setting `.classes()`, they might not yield the expected result due to other `.style()` settings, which might need to be removed or replaced using the `remove=` or `replace=` attributes.

**Comment by clueple** [comment]
```py
df = pd.DataFrame(data={'col1': [1, 2], 'col2': [3, 4]})

table = ui.table({})
table.view.load_pandas_frame(df)
```

Thanks for the answer but ui.table has no attribute 'view' .
Is there any user-friendly way to show a DataFrame?

Or can we `pd.DataFrame(whatever_data).to_dict('split')` and then 
```py
ui.table({
  "columnDefs":   [point_to_columns_array],
  "rowdata": [point_to_rows_array]
})
```
?

It'd be nice if there are more details in the documentation.

Thanks in advance for any ideas.

**Reply by phoskee** [reply]
Try `table.options['rowData'] = [point_to_rows_array]`

**Reply by clueple** [reply]
Thanks for your hint.

That's the result of my code.  Still figuring out how to fix it.
![nicegui_table_fail](https://user-images.githubusercontent.com/19367464/220820220-4fc4ed81-a5e2-48ac-b2ee-5f68b0803a6b.PNG)

**Reply by phoskee** [reply]
Next time copy/paste your code here!
Btw, you forgot the column definition

```py
ui.table({
    'columnDefs': [
        {'headerName': 'your stuff', 'field': 'your stuff'},
    ],
    'rowData': [],
})
```

**Reply by clueple** [reply]
> Next time copy/paste your code here! Btw, you forgot the column definition
> 
> ```python
> ui.table({
>     'columnDefs': [
>         {'headerName': 'your stuff', 'field': 'your stuff'},
>     ],
>     'rowData': [],
> })
> ```

Thanks for reminding, I figured that out (columnDefs equals = [ list comprehension of each item in the df.columns as a dictionary]

**Comment by falkoschindler** [comment, marked as answer]
You're right, @clueple! This Q&A was outdated. Since NiceGUI 1.0 there is no JustPy `view` anymore.

Here is a snippet on how to create a `ui.table` from a Pandas DataFrame:

```py
df = pd.DataFrame(data={'col1': [1, 2], 'col2': [3, 4]})

table = ui.table({
    'columnDefs': [{'field': col} for col in df.columns],
    'rowData': df.to_dict('records'),
})
```

**Reply by clueple** [reply]
> You're right, @clueple! This Q&A was outdated. Since NiceGUI 1.0 there is no JustPy `view` anymore.
> 
> Here is a snippet on how to create a `ui.table` from a Pandas DataFrame:
> 
> ```python
> df = pd.DataFrame(data={'col1': [1, 2], 'col2': [3, 4]})
> 
> table = ui.table({
>     'columnDefs': [{'field': col} for col in df.columns],
>     'rowData': df.to_dict('records'),
> })
> ```

Thanks for the detail code.  What I did was:

```py
#create a ui.table object
table = ui.table({}) 

# assign the 'columnDefs' to be a list comprehension of each df column item following the API's dictionary format
table.options['columnDefs'] = [ { f"headerName: {item.upper()}, 'field':  {item} "    }       for idx, item in enumerate(df.columns)]

# assign the 'rowData' to be the array of of df's data 
table.options['rowData'] = df.to_dict('records')
```

I'm trying to simplify the codes with a data class.  Will share when it's done.

Grazie.

**Comment by bdshoener** [comment]
I know this is closed, but with the switch of `ui.table` to QTable, the syntax is a little different now:

```py
ui.table(
    columns=[{'field': col} for col in df.columns],
    rows=df.to_dict('records'),
)
```

**Reply by falkoschindler** [reply]
That's right, @bdshoener!
For column headers you should also set `name` and `label`:
```py
df = pd.DataFrame(data={'col1': [1, 2], 'col2': [3, 4]})
ui.table(
    columns=[{'name': col, 'label': col, 'field': col} for col in df.columns],
    rows=df.to_dict('records'),
)
```
I'll add a demo to the documentation.

**Reply by clueple** [reply]
> I know this is closed, but with the switch of `ui.table` to QTable, the syntax is a little different now:
> 
> ```python
> ui.table(
>     columns=[{'field': col} for col in df.columns],
>     rows=df.to_dict('records'),
> )
> ```

I've come up with one (see if it helps)
```py
from nicegui import ui
import pandas as pd

file = r'g:/pybkp/chartOfAcct.csv'  #change the file location to suit your need

df = pd.read_csv(file)  #store the 'file' as a dataframe object

def get_df_col(df):
    """ store the columns and its headerName as the upper case of the field name in a list"""
    return [{"headerName": col.upper(), "field": col} for col in df.columns]

def get_df_row(df):
    """  return the data of your dataframe  """
    return df.to_dict('records')

#read the dataframe as a ui.aggrid object
ui.aggrid({
    'columnDefs': get_df_col(df=df),
    'rowData': get_df_row(df=df) 
}).classes('max-h-40')

ui.label('This is testing')
ui.button('123')

ui.run()
```
